src/main/java/io/zksync/wallet/DefaultZkASyncWallet.java

```diff
@@ -0,0 +1,301 @@
+        return this.getPubKeyHash()
+            .thenApply(pubKey -> Objects.equals(pubKey, this.zkSigner.getPublicKeyHash()));
+    }
+
+    @Override
+    public CompletableFuture<AccountState> getState() {
+        return this.provider.getState(this.getAddress());
+    }
+
+    @Override
+    public AsyncProvider getProvider() {
+        return this.provider;
+    }
+
+    @Override
+    public CompletableFuture<String> getPubKeyHash() {
+        if (this.pubKeyHash == null) {
+            return getState()
+                .thenApply(this::setAccountInfo)
+                .thenApply(state -> state.getCommitted().getPubKeyHash());
+        } else {
+            return CompletableFuture.completedFuture(this.pubKeyHash);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Integer> getAccountId() {
+        if (this.accountId == null) {
+            return getState()
+                .thenApply(this::setAccountInfo)
+                .thenApply(AccountState::getId);
+        } else {
+            return CompletableFuture.completedFuture(this.accountId);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Tokens> getTokens() {
+        if (this.tokens == null) {
+            return this.provider.getTokens()
+                .thenApply(tokens -> {
+                    this.tokens = tokens;
+                    return tokens;
+                });
+        } else {
+            return CompletableFuture.completedFuture(this.tokens);
+        }
+    }
+
+    @Override
+    public <T extends ZkSyncTransaction> CompletableFuture<String> submitTransaction(SignedTransaction<T> transaction) {
+        return submitSignedTransaction(transaction.getTransaction(), transaction.getEthereumSignature());
+    }
+
+    @Override
+    public String getAddress() {
+        return this.ethSigner.getAddress();
+    }
+
+    @Override
+    public CompletableFuture<Integer> getNonce() {
+        return getState()
+            .thenApply(state -> state.getCommitted().getNonce());
+    }
+
+    @Override
+    public EthereumProvider createEthereumProvider(Web3j web3j, ContractGasProvider contractGasProvider) {
+        String contractAddress = this.provider.contractAddress().join().getMainContract();
+        ZkSync contract = ZkSync.load(contractAddress, web3j, this.ethSigner.getTransactionManager(), contractGasProvider);
+        DefaultEthereumProvider ethereum = new DefaultEthereumProvider(web3j, this.ethSigner, contract);
+        return ethereum;
+    }
+
+    private CompletableFuture<String> submitSignedTransaction(ZkSyncTransaction signedTransaction,
+                                         EthSignature ethereumSignature,
+                                         boolean fastProcessing) {
+        return provider.submitTx(signedTransaction, ethereumSignature, fastProcessing);
+    }
+
+    private CompletableFuture<String> submitSignedTransaction(ZkSyncTransaction signedTransaction,
```

Comment:
Submitting should be public as well as signing in the Default Wallet

Reply:
Do you mean that wallet should have method for send unsigned transactions (means sign it inside wallet)?

Reply:
I mean that functions like `submitSignedTransaction` or `signTransaction` (or whatever) should be public so that the users can separate the submission and the signing process